vmware-tanzu/nsx-operator
# Unify NSX resoruce ID and name connector as underline

Previsouly, we use hyphen "-" to connect strings when building NSX resource name, and use underline "_" to connect strings when builindg NSX resource ID.

This patch is to unify NSX resource ID and name connecotr as underline when building ID and name from K8s CR.

For the NSX resoruce ID and name convention,
this patch is follow the standard in PR:https://github.com/vmware-tanzu/nsx-operator/pull/643

These NSX resources name are impacted in this change:
VPC
Subnet
SubnetPort
SecurityPolicy and NetworkPolicy
NSGroup and IPSetGroup
NSRule
Share
StaticRoute
IPAllocation

This patch is aslo to add two flags for GC and vpcCleanup in order to
distinguish these two cases.

Test Done:
1. Create ingress Networkpolicy
```
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: np-app-access-ing
  namespace: ns-2
  annotations:
    ncp/log_traffic: "OFF"
spec:
  podSelector:
    matchLabels:
      role: db
  ingress:
  - from:
    - podSelector:
        matchLabels:
          app: coffee
    - namespaceSelector:
        matchLabels:
          ns-name: ns-3
      podSelector:
        matchLabels:
          app: tea
    - ipBlock:
        cidr: 172.26.0.8/29
        except:
        - 172.26.0.11/32
    ports:
    - protocol: TCP
      port: 6001
    - protocol: TCP
      port: 7000
      endPort: 7010
```

The generated ingress allow SecurityPolicy:
```
SecurityPolicy name: np-app-access
SecurityPolicy ID: np-app-access_d9dabc6b-9668-4cd1-9c52-264ff60be3f8_allow
```
Rule:
```
Name: TCP.6001_ingress_allow
ID: np-app-access_d9dabc6b-9668-4cd1-9c52-264ff60be3f8_allow_0_6c2a026ca143812daa72699fb924ee36b33b5cdc_0_0
```

The generated ingress isolation SecurityPolicy:
```
SecurityPolicy name:np-app-access
SecurityPolicy ID: np-app-access_d9dabc6b-9668-4cd1-9c52-264ff60be3f8_isolation
```
Rule:
```
Name: ingress_isolation
ID:np-app-access_d9dabc6b-9668-4cd1-9c52-264ff60be3f8_isolation_0_114fed106ef3b5eae2a583f312435e84c02ca97f_0_0
```

Project Share
```
name: dy-project1_group_np-app-access_0_src_share 
ID: dy-project1_group_np-app-access_d9dabc6b-9668-4cd1-9c52-264ff60be3f8_allow_0_src_share
```

Project Group
```
Name: np-app-access_0_src
ID: np-app-access_d9dabc6b-9668-4cd1-9c52-264ff60be3f8_allow_0_src
```

VPC NSGroup:
```
Name: test_np-app-access_scope
ID: np-app-access_d9dabc6b-9668-4cd1-9c52-264ff60be3f8_isolation_scope
```

2. Create/delete SecurityPolicy in VPC mode.

3.  Create/delete SecurityPolicy with namedport
The generated IPSet group:
```
Name: TCP.db-port.TCP.3366_egress_allow_ipset
ID: sp-namedport-web_90fc82ff-395e-47d0-bd02-534d71b2e434_2_6227e9f64a8ca9ab0e34891f3725a4357bc8d917_0_1_ipset
Name: TCP.db-port.TCP.80_egress_allow_ipset
ID: sp-namedport-web_90fc82ff-395e-47d0-bd02-534d71b2e434_2_6227e9f64a8ca9ab0e34891f3725a4357bc8d917_0_0_ipset
```

4. T1 upgrade case:
Create a SecurityPolicy in v4.2.1 code,

![image](https://github.com/user-attachments/assets/c69f2987-eb98-42d1-9015-171df048121c)

and start nsx operator with this patch to see if name are changed from "-" to "_".

![image](https://github.com/user-attachments/assets/b7a2e23f-02a2-4b7e-91cc-d2cdd2456403)

5. Create SecurityPolicy in T1 and do GC

6. Create SecurityPolicy in VPC and do GC

7. Create NetworkPolicy in VPC and do GC
